File: backend/src/main/java/de/interaapps/pastefy/controller/analytics/AnalyticsController.java
package de.interaapps.pastefy.controller.analytics;

import de.interaapps.pastefy.Pastefy;
import de.interaapps.pastefy.analytics.AnalyticsQuery;
import de.interaapps.pastefy.analytics.AnalyticsResponse;
import de.interaapps.pastefy.controller.HttpController;
import de.interaapps.pastefy.exceptions.FeatureDisabledException;
import de.interaapps.pastefy.exceptions.PermissionsDeniedException;
import de.interaapps.pastefy.model.database.Paste;
import de.interaapps.pastefy.model.database.AuthKey;
import de.interaapps.pastefy.model.database.User;
import org.javawebstack.http.router.Exchange;
import org.javawebstack.http.router.router.annotation.PathPrefix;
import org.javawebstack.http.router.router.annotation.With;
import org.javawebstack.http.router.router.annotation.params.Attrib;
import org.javawebstack.http.router.router.annotation.params.Path;
import org.javawebstack.http.router.router.annotation.verbs.Get;

import java.util.Objects;

@PathPrefix("/api/v2/analytics")
@With("auth")
public class AnalyticsController extends HttpController {
    @Get("/admin")
    @With("admin")
    public AnalyticsResponse admin(Exchange exchange, @Attrib("authkey") AuthKey authKey) {
        authKey.checkPermission("analytics:read");
        return query(exchange);
    }

    @Get("/pastes/{id}")
    public AnalyticsResponse paste(Exchange exchange, @Path("id") String id, @Attrib("user") User user, @Attrib("authkey") AuthKey authKey) {
        authKey.checkPermission("analytics:read");
        Paste paste = Paste.get(id);
        if (paste == null || (!user.isAdmin() && !Objects.equals(user.getId(), paste.getUserId()))) {
            throw new PermissionsDeniedException();
        }
        AnalyticsQuery query = AnalyticsQuery.from(exchange);
        query.filters.put("paste_key", paste.getKey());
        return query(query);
    }

    @Get("/user")
    public AnalyticsResponse user(Exchange exchange, @Attrib("user") User user, @Attrib("authkey") AuthKey authKey) {
        authKey.checkPermission("analytics:read");
        AnalyticsQuery query = AnalyticsQuery.from(exchange);
        query.filters.put("paste_user_id", user.getId());
        return query(query);
    }

    private AnalyticsResponse query(Exchange exchange) {
        return query(AnalyticsQuery.from(exchange));
    }

    private AnalyticsResponse query(AnalyticsQuery query) {
        if (!Pastefy.getInstance().analyticsEnabled()) throw new FeatureDisabledException();
        return Pastefy.getInstance().getAnalyticsService().query(query);
    }
}

File: backend-kt/src/main/kotlin/de/interaapps/pastefy/controller/analytics/AnalyticsController.kt
package de.interaapps.pastefy.controller.analytics

import de.interaapps.pastefy.auth.annotations.*
import de.interaapps.pastefy.dto.analytics.AnalyticsResponse
import de.interaapps.pastefy.dto.analytics.AnalyticsQuery
import de.interaapps.pastefy.entities.AuthKey
import de.interaapps.pastefy.entities.User
import jakarta.servlet.http.HttpServletRequest
import de.interaapps.pastefy.exceptions.FeatureDisabledException
import de.interaapps.pastefy.exceptions.PermissionsDeniedException
import de.interaapps.pastefy.infrastructure.analytics.AnalyticsService
import de.interaapps.pastefy.service.PasteService
import org.springframework.beans.factory.ObjectProvider
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/v2/analytics")
@Authenticated
class AnalyticsController(
    private val analyticsProvider: ObjectProvider<AnalyticsService>,
    private val pasteService: PasteService,
) {
    @GetMapping("/admin")
    @AdminRoute
    @RequiresPermission("analytics:read")
    fun admin(request: HttpServletRequest, @CurrentAuthKey authKey: AuthKey): AnalyticsResponse = query(AnalyticsQuery.from(request))

    @GetMapping("/pastes/{id}")
    @RequiresPermission("analytics:read")
    fun paste(@PathVariable id: String, request: HttpServletRequest, @CurrentUser user: User, @CurrentAuthKey authKey: AuthKey): AnalyticsResponse {
        val paste = pasteService.get(id)
        if (paste == null || (!user.isAdmin && user.id != paste.userId)) throw PermissionsDeniedException()
        return query(AnalyticsQuery.from(request).apply { filters["paste_key"] = paste.key })
    }

    @GetMapping("/user")
    @RequiresPermission("analytics:read")
    fun user(request: HttpServletRequest, @CurrentUser user: User, @CurrentAuthKey authKey: AuthKey): AnalyticsResponse =
        query(AnalyticsQuery.from(request).apply { filters["paste_user_id"] = user.id })

    private fun query(query: AnalyticsQuery): AnalyticsResponse =
        (analyticsProvider.ifAvailable ?: throw FeatureDisabledException()).query(query)
}
